### ISSUE TITLE
Dataclasses not detected outside of direct @dataclass usage

### ISSUE THREAD

**Issue by Terrance** [microsoft/pyright#1754]
## Describe the bug

Dataclass validation is absent when `@dataclass` is not directly applied to the class.  One example use case for this is proxying the `@dataclass(...)` call behind another decorator in order to provide common options to many classes.

## To reproduce

```python
from dataclasses import dataclass

@dataclass(init=True)
class A:
    optional: int = 1
    required: int  # Dataclass fields without default values cannot appear after fields with default values

def deco(cls):
    return dataclass(init=True)(cls)

@deco
class B:
    optional: int = 1
    required: int
    # No errors reported?
```

`A` and `B` should be the same structure, but the check for required arguments after optional ones doesn't happen on `B`.

At a guess, this is because dataclasses themselves are not represented by types, just detected from the `@dataclass` syntax.

## VS Code extension or command-line

Pyright 1.1.130 in VSCode.

**Comment by erictraut** [comment]
The "dataclass" decorator is a complex transform, and a static type checker needs to understand the details of the class it is being applied to. It cannot apply such a transform to a generic type or a type that is passed dynamically to a function, like in your `deco` example above. Your example above doesn't work with mypy either, for good reason.

**Comment by Terrance** [comment]
Fair enough, I suspected this would probably be out of scope.
